text/0000-swap-out-jemalloc.md

```diff
@@ -0,0 +1,235 @@
+
+The compiler will also not immediately verify that crates tagged with
+`#![allocator]` do indeed define an appropriate allocation API, and vice versa
+if a crate defines an allocation API the compiler will not verify that it is
+tagged with `#![allocator]`. This means that the only meaning `#![allocator]`
+has to the compiler is to signal that the default allocator should not be
+linked.
+
+### Default allocator specifications
+
+Target specifications will be extended with two keys: `lib_allocation_crate`
+and `exe_allocation_crate`, describing the default allocator crate for these
+two kinds of artifacts for each target. The compiler will by default have all
+targets redirect to `alloc_system` for both scenarios, but `alloc_jemalloc` will
+be used for binaries on OSX, Bitrig, DragonFly, FreeBSD, Linux, OpenBSD, and GNU
+Windows. MSVC will notably **not** use jemalloc by default for binaries (we
+don't currently build jemalloc on MSVC).
+
+### Injecting an allocator
+
+As described above, the compiler will inject an allocator if necessary into the
+current compilation. The compiler, however, cannot blindly do so as it can
+easily lead to link errors (or worse, two allocators), so it will have some
+heuristics for only injecting an allocator when necessary. The steps taken by
+the compiler for any particular compilation will be:
+
+* If no crate in the dependency graph is tagged with `#![needs_allocator]`, then
+  the compiler does not inject an allocator.
+* If only an rlib is being produced, no allocator is injected.
+* If any crate tagged with `#[allocator]` has been explicitly linked to (e.g.
+  via an `extern crate` statement directly or transitively) then no allocator is
+  injected.
+* If two allocators have been linked to explicitly an error is generated.
+* If only a binary is being produced, then the target's `exe_allocation_crate`
+  value is injected, otherwise the `lib_allocation_crate` is injected.
+
+The compiler will also record that the injected crate is injected, so later
+compilations know that rlibs don't actually require the injected crate at
+runtime (allowing it to be overridden).
+
+### Allocators in practice
+
+Most libraries written in Rust wouldn't interact with the scheme proposed in
+this RFC at all as they wouldn't explicitly link with an allocator and generally
+are compiled as rlibs. If a Rust dynamic library is used as a dependency, then
+its original choice of allocator is propagated throughout the crate graph, but
+this rarely happens (except for the compiler itself, which will continue to use
+jemalloc).
+
+Authors of crates which are embedded into other runtimes will start using the
+system allocator by default with no extra annotation needed. If they wish to
+funnel Rust allocations to the same source as the host application's allocations
+then a crate can be written and linked in.
+
+Finally, providers of allocators will simply provide a crate to do so, and then
```

Comment:
Unfortunately this RFC doesn't currently easily allow this sort of instrumentation to happen. If we wanted to support this right out of the gate, this RFC would necessitate four crates:

* Two crates for implementing the allocation API, but not tagged with `#![allocator]`. There'd be one crate for jemalloc and one for the system.
* Two crates for linking to the previous crates, but are tagged with `#![allocator]` and redirect the formal allocation API into the desired crate.

In a nutshell, if you want to write an allocator which can be instrumented, or shimmed then you need to write a crate which is **not** tagged `#![allocator]` but probably still exposes the allocation API via normal Rust functions. The provider of the allocator would then write their own shims that redirect to the allocator desired after the instrumentation has happened.

Does that make sense? If so I'll add some words.